app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt

```diff
@@ -0,0 +1,178 @@
+        addViewModel.pinErrorMsg.set(activity.resources.getString(R.string.add_profile_error_pin_length))
+        failed = true
+      }
+      if (pin != confirmPin) {
+        addViewModel.confirmPinErrorMsg.set(activity.resources.getString(R.string.add_profile_error_pin_confirm_wrong))
+        failed = true
+      }
+      if (failed)  {
+        binding.scroll.smoothScrollTo(0,0)
+        return@setOnClickListener
+      }
+      profileManagementController.addProfile(name, pin, selectedImage, allowDownloadAccess, isAdmin = false).observe(activity, Observer {
+        if (it.isSuccess()) {
+          val intent = Intent(activity, ProfileActivity::class.java)
+          intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+          activity.startActivity(intent)
+        } else if (it.isFailure()) {
+          when (it.getErrorOrNull()) {
+            is ProfileManagementController.ProfileNameNotUniqueException -> addViewModel.nameErrorMsg.set(activity.resources.getString(R.string.add_profile_error_name_not_unique))
+            is ProfileManagementController.ProfileNameOnlyLettersException -> addViewModel.nameErrorMsg.set(activity.resources.getString(R.string.add_profile_error_name_only_letters))
+          }
+          binding.scroll.smoothScrollTo(0,0)
+        }
+      })
+    }
+  }
+
+  fun handleOnActivityResult(data: Intent?) {
+    data?.let {
+      selectedImage = data.data
+      Glide.with(activity)
+        .load(selectedImage)
+        .centerCrop()
+        .apply(RequestOptions.circleCropTransform())
+        .into(uploadImageView)
+    }
+  }
+
+  private fun addTextChangeListeners(binding: AddProfileActivityBinding) {
+    fun setValidPin() {
+      if (inputtedPin && inputtedConfirmPin) {
+        addViewModel.validPin.set(true)
+      } else {
+        binding.allowDownloadSwitch.isChecked = false
+        addViewModel.validPin.set(false)
+      }
+    }
+
+    binding.inputPin.addTextChangedListener(object: TextWatcher {
+      override fun onTextChanged(pin: CharSequence?, start: Int, before: Int, count: Int) {
+        pin?.let {
+          addViewModel.pinErrorMsg.set("")
+          inputtedPin = it.isNotEmpty()
+          setValidPin()
+        }
+      }
+      override fun afterTextChanged(confirmPin: Editable?) {}
+      override fun beforeTextChanged(p0: CharSequence?, start: Int, count: Int, after: Int) {}
+    })
+
+    binding.inputConfirmPin.addTextChangedListener(object: TextWatcher {
+      override fun onTextChanged(confirmPin: CharSequence?, start: Int, before: Int, count: Int) {
+        confirmPin?.let {
+          addViewModel.confirmPinErrorMsg.set("")
+          inputtedConfirmPin = confirmPin.isNotEmpty()
+          setValidPin()
+        }
+      }
+      override fun afterTextChanged(confirmPin: Editable?) {}
+      override fun beforeTextChanged(p0: CharSequence?, start: Int, count: Int, after: Int) {}
+    })
+
+    binding.inputName.addTextChangedListener(object: TextWatcher {
```

Comment:
These text watchers seem really repetitive. Could we factor them into a common helper that takes a lambda that performs the actions we want? Then it might be a bit simpler here, e.g.:

```kotlin
addTextChangedListener(binding.inputName) {
  confirmPin?.let { ... }
}
```